src/document-click.js

```diff
@@ -113,18 +145,37 @@ export class ClickHandler {
  * @param {!Event} e
  * @param {!./service/viewport-impl.Viewport} viewport
  * @param {!./service/history-impl.History} history
+ * @param {!./service/url-replacements-impl.UrlReplacements} urlReplacements
  * @param {boolean} isIosSafari
+ * @param {boolean} isIframed
+ * @param {boolean} isCapture whether event was caught during capture
+ *                  propagation.
  */
-export function onDocumentElementClick_(e, viewport, history, isIosSafari) {
+export function onDocumentElementClick_(e, viewport, history, urlReplacements,
+                                        isIosSafari, isCapture) {
   if (e.defaultPrevented) {
     return;
   }
 
-  const target = closestByTag(e.target, 'A');
+  // If within a shadowRoot, the event target will be the host element due to
+  // event target rewrite.  Given that it is possible a shadowRoot could be
+  // within an anchor tag, we need to check the event path prior to looking
+  // at the host element's closest tags.
+  const target = getElementByTagNameFromEventShadowDomPath_(e, 'A') ||
+      closestByTag(e.target, 'A');
   if (!target) {
     return;
   }
 
+  // Capture event listener only interested in anchor target href expansion.
+  if (isCapture) {
```

Comment:
This should definitely be its own function, eliminating [this](https://github.com/ampproject/amphtml/pull/4773/files#r77050396) whole confusion.